src/compaction_filter.rs

```diff
@@ -81,6 +114,193 @@ pub unsafe fn new_compaction_filter(
         filter,
         name,
     );
-    crocksdb_ffi::crocksdb_compactionfilter_set_ignore_snapshots(filter, ignore_snapshots);
-    Ok(CompactionFilterHandle { inner: filter })
+    // In latest rocksdb `ignore_snapshots` will be always true.
+    // TODO: remove it after rocksdb is upgraded.
+    crocksdb_ffi::crocksdb_compactionfilter_set_ignore_snapshots(filter, true);
+    filter
+}
+
+pub struct CompactionFilterContext(DBCompactionFilterContext);
+impl CompactionFilterContext {
+    pub fn is_full_compaction(&self) -> bool {
+        let ctx = &self.0 as *const DBCompactionFilterContext;
+        unsafe { crocksdb_ffi::crocksdb_compactionfiltercontext_is_full_compaction(ctx) }
+    }
+
+    pub fn is_manual_compaction(&self) -> bool {
+        let ctx = &self.0 as *const DBCompactionFilterContext;
+        unsafe { crocksdb_ffi::crocksdb_compactionfiltercontext_is_manual_compaction(ctx) }
+    }
+}
+
+pub trait CompactionFilterFactory {
+    fn create_compaction_filter(
+        &self,
+        context: &CompactionFilterContext,
+    ) -> *mut DBCompactionFilter;
+}
+
+#[repr(C)]
+struct CompactionFilterFactoryProxy {
+    name: CString,
+    factory: Box<dyn CompactionFilterFactory>,
```

Comment:
Do we need to access the `CompactionFilterFactory` instance on tikv side, or will the `CompactionFilterFactory` instance hold resource in tikv so it can obtain safe-point, etc? If it is the former, holding a `Arc<dyn CompactionFilterFactory>` here is better.

Reply:
In TiKV we don't need to access the factory.

Reply:
👌